### PR TITLE
[Optimizer] A slightly more efficient version of lamb

### PR DESCRIPTION
**Pull request by eric-haibin-lin**
## Description ##
Removed control flow in lamb, which causes python code to synchronize. 
Applied some inplace operation to reduce memory footprint. 
@kenjewu  @apeforest 

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [x] Code is well-documented

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by mli** [comment]
Job PR-929/1 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-929/1/index.html

**Comment by mli** [comment]
Job PR-929/2 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-929/2/index.html

**Comment by mli** [comment]
Job PR-929/3 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-929/3/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 1]
```diff
@@ -104,25 +104,34 @@ def update(self, index, weight, grad, state):
...
+            g = mean_hat
 
         r2 = g.norm()
 
         # calculate lamb_trust_ratio
-        r = 1. if r1 == 0. or r2 == 0. else r1 / r2
+        ratio = r1 / r2
+        # becomes NaN if ratio == NaN or 0
+        nan_or_zero = ratio / ratio
+        r = where(nan_or_zero, ones_like(ratio), ratio)
```
Comment (szhengac) [inline src/gluonnlp/optimizer/lamb.py]: if ratio is not either nan or 0, `where` will return 1, while we want it to return ratio.

Comment (eric-haibin-lin) [reply src/gluonnlp/optimizer/lamb.py]: My bad. The variable name is wrong. Should be nan_or_one

**Comment by mli** [comment]
Job PR-929/4 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-929/4/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 2]
```diff
@@ -127,9 +127,9 @@ def update(self, index, weight, grad, state):
 
         # calculate lamb_trust_ratio
         ratio = r1 / r2
-        # becomes NaN if ratio == NaN or 0
-        nan_or_zero = ratio / ratio
-        r = where(nan_or_zero, ones_like(ratio), ratio)
+        # becomes NaN if ratio == NaN or 0, otherwise 1
+        nan_or_one = ratio / ratio
+        r = where(nan_or_one, ones_like(ratio), ratio)
```
Comment (szhengac) [inline src/gluonnlp/optimizer/lamb.py]: I mean if nan_or_one is 1 (True), this line returns 1, but we want it to return ratio (False).

**Comment by apeforest** [comment]
@anirudhacharya Please help to review.

File: src/gluonnlp/optimizer/lamb.py (hunk 1 above)
Comment (eric-haibin-lin) [reply src/gluonnlp/optimizer/lamb.py]: Thanks.. I fixed the condition with NaN now: 
```
>>> mx.nd.where(1-nan, mx.nd.array([1],ctx=mx.gpu()), mx.nd.array([0.2],ctx=mx.gpu()))

[1.]
<NDArray 1 @gpu(0)>
>>> mx.nd.where(1-mx.nd.array([1],ctx=mx.gpu()), mx.nd.array([1],ctx=mx.gpu()), mx.nd.array([0.2],ctx=mx.gpu()))

[0.2]
<NDArray 1 @gpu(0)>
```

**Comment by mli** [comment]
Job PR-929/5 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-929/5/index.html

File: src/gluonnlp/optimizer/lamb.py [hunk 3]
```diff
@@ -104,25 +104,34 @@ def update(self, index, weight, grad, state):
...
+            mean_hat += wd * weight
+            g = mean_hat
 
         r2 = g.norm()
 
         # calculate lamb_trust_ratio
-        r = 1. if r1 == 0. or r2 == 0. else r1 / r2
+        ratio = r1 / r2
+        # becomes NaN if ratio == NaN or 0, otherwise 0
+        nan_or_zero = 1 - ratio / ratio
```
Comment (anirudhacharya) [inline src/gluonnlp/optimizer/lamb.py]: `nan_or_zero` will throw a `ZeroDivisionError` when `ratio` is equal to 0

**Comment by szhengac** [comment]
Will NAN trigger error message?

**Comment by eric-haibin-lin** [comment]
> Will NAN trigger error message?

Not at the moment if we execute it using the asynchronous engine...

**Pull request closed without merging**